### PR TITLE
SONAR-9739 add ws api/system/health

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/health/HealthCheckerImpl.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public Health apply(Health left, Health right) {
+      Health.Builder builder = newHealthCheckBuilder(highestOf(left.getStatus(), right.getStatus()));
+      left.getCauses().forEach(builder::addCause);
+      right.getCauses().forEach(builder::addCause);
+      return builder.build();
+    }
+
+    private Health.Status highestOf(Health.Status left, Health.Status right) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/health/HealthCheckerImpl.java]: MAJOR Make "highestOf" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/health/WebServerStatusCheck.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.sonar.server.platform.Platform;
+import org.sonar.server.platform.db.migration.DatabaseMigrationState;
+
+import static org.sonar.server.health.Health.newHealthCheckBuilder;
+
+/**
+ * A HealthCheck for the running status of the web server.
+ */
+public class WebServerStatusCheck implements HealthCheck {
+  private final EnumSet<DatabaseMigrationState.Status> VALID_DATABASEMIGRATION_STATUSES = EnumSet.of(
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/health/WebServerStatusCheck.java]: MINOR Rename this field "VALID_DATABASEMIGRATION_STATUSES" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00116)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/platform/ws/AbstractHealthAction.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import com.google.common.io.Resources;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.server.health.Health;
+import org.sonar.server.health.HealthChecker;
+import org.sonar.server.ws.WsUtils;
+import org.sonarqube.ws.WsSystem;
+
+public abstract class AbstractHealthAction implements SystemWsAction {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/platform/ws/AbstractHealthAction.java]: Using class inheritance for checking a state (safe mode or not) is too complex and hard to maintain. The web service should be able to request the state through an injected component (something like `lifecycle.isSafeMode() || isSystemAdministrator()`).

File: server/sonar-server/src/main/java/org/sonar/server/platform/ws/AbstractHealthAction.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    controller.createAction("health")
+      .setDescription("Provide health status of the current SonarQube instance." +
+        "<p>status: the health status" +
+        " <ul>" +
+        " <li>GREEN: SonarQube is fully operational</li>" +
+        " <li>YELLOW: SonarQube is operational but something must be fixed to be fully operational</li>" +
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/platform/ws/AbstractHealthAction.java]: What is the difference between "operational" and "fully operational" ? I suggest to drop the concept of "fully operational" and to mention the risk of becoming red if warnings are not fixed.
cc @christophelevis

File: server/sonar-server/src/main/resources/org/sonar/server/platform/ws/example-health.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "health": "YELLOW",
+  "causes": [
+    {
+      "message": "Elasticsearch's status is YELLOW"
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/resources/org/sonar/server/platform/ws/example-health.json]: Isn't it "Elasticsearch status" without "'s" ?

File: server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+import static org.apache.commons.lang.RandomStringUtils.randomAlphanumeric;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.sonar.server.health.Health.newHealthCheckBuilder;
+
+public class HealthTest {
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  private final Random random = new Random();
+  private final Health.Status randomStatus = Health.Status.values()[random.nextInt(Health.Status.values().length)];
```
Comment (simonbrandhof) [inline server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java]: I consider randomized test as an anti-pattern. It's impossible to reproduce potential errors and it brings complexity to read and maintain the test.

File: server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    });
+
+    assertThat(builder1.build().hashCode())
+      .isEqualTo(builder1.build().hashCode())
+      .isEqualTo(builder2.build().hashCode())
+      .isEqualTo(builder2.build().hashCode());
+  }
+
+  @Test
+  public void verify_toString() {
```
Comment (simonbrandhof) [inline server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java]: This test is over-complex regarding the value and the code being tested. It's just testing a trivial toString() method. Testing a single state is far enough, for example GREEN and 2 messages. No need to test all the cases (0, 1 or more causes).

File: sonar-ws/src/main/protobuf/ws-system.proto [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import "ws-commons.proto";
+
+option java_package = "org.sonarqube.ws";
+option java_outer_classname = "WsSystem";
+option optimize_for = SPEED;
+
+// POST api/ce/health
+message HealthResponse {
+  optional HealthStatus health = 1;
```
Comment (simonbrandhof) [inline sonar-ws/src/main/protobuf/ws-system.proto]: I suggest to name the enum `Health`.

File: sonar-ws/src/main/protobuf/ws-system.proto [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+package sonarqube.ws.system;
+
+import "ws-commons.proto";
+
+option java_package = "org.sonarqube.ws";
+option java_outer_classname = "WsSystem";
+option optimize_for = SPEED;
+
+// POST api/ce/health
```
Comment (simonbrandhof) [inline sonar-ws/src/main/protobuf/ws-system.proto]: should be GET api/system/health

File: sonar-ws/src/main/protobuf/ws-system.proto [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program; if not, write to the Free Software Foundation,
+// Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+syntax = "proto2";
+
+package sonarqube.ws.system;
+
+import "ws-commons.proto";
```
Comment (simonbrandhof) [inline sonar-ws/src/main/protobuf/ws-system.proto]: this import can be removed

File: server/sonar-server/src/main/java/org/sonar/server/platform/ws/AbstractHealthAction.java (hunk 3 above)
Comment (sns-seb) [reply server/sonar-server/src/main/java/org/sonar/server/platform/ws/AbstractHealthAction.java]: as discussed offline, I will keep things as is

File: server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java (hunk 6 above)
Comment (sns-seb) [reply server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java]: I'll rename this to anyStatus.
Reproducibility problem is handled by adding description to assertions which provide information about which values where used when the test failed

File: server/sonar-server/src/main/resources/org/sonar/server/platform/ws/example-health.json (hunk 5 above)
Comment (sns-seb) [reply server/sonar-server/src/main/resources/org/sonar/server/platform/ws/example-health.json]: fixed

File: server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java (hunk 7 above)
Comment (sns-seb) [reply server/sonar-server/src/test/java/org/sonar/server/health/HealthTest.java]: problem with 2 messages is that causes are stored in a set and order in a set is not predictive (even in the same JVM, I got the error, it surprised me)

File: sonar-ws/src/main/protobuf/ws-system.proto (hunk 10 above)
Comment (sns-seb) [reply sonar-ws/src/main/protobuf/ws-system.proto]: done

File: sonar-ws/src/main/protobuf/ws-system.proto (hunk 9 above)
Comment (sns-seb) [reply sonar-ws/src/main/protobuf/ws-system.proto]: good catch

File: sonar-ws/src/main/protobuf/ws-system.proto (hunk 8 above)
Comment (sns-seb) [reply sonar-ws/src/main/protobuf/ws-system.proto]: done

**Pull request merged**